draft-gpew-priv-ppm.md

```diff
@@ -1004,21 +1125,139 @@ fetch certificates: https://datatracker.ietf.org/doc/html/rfc8555#section-7.4.2]
+the hash values with a bitwise-XOR operation.
+
+Then, for each aggregator endpoint `[aggregator]` in the parameters associated
+with `CollectReq.task_id` (see {{collect-flow}}) except its own, the leader sends
+a POST request to `[aggregator]/aggregate_share` with the following message:
+
+~~~
+struct {
+  TaskID task_id;
+  Interval batch_interval;
+  uint64 report_count;
+  opaque checksum[32];
+  opaque helper_state<0..2^16>;
+} AggregateShareReq;
+~~~
+
+* `task_id` is the task ID associated with the PPM parameters.
+* `batch_interval` is the batch interval of the request.
+* `report_count` is the number of reports included in the aggregation.
+* `checksum` is the checksum computed over the set of client reports.
+* `helper_state` is the helper's state, which is carried across requests from
+  the leader.
+
+To handle the leader's request, the helper first ensures that the request meets
+the requirements for batch parameters following the procedure in
+{{batch-parameter-validation}}. If the batch parameters are invalid, then it
+MUST abort with error "batchInvalid".
+
+Next, it computes a checksum based on its view of the output shares included in
+the batch window, and checks that the `report_count` and `checksum` included in
+the request match its computed values. If not, then it MUST abort with error
+"batchMismatch".
+
+Next, it computes the aggregate share `agg_share` corresponding to the set of
+output shares, denoted `out_shares`, for the batch interval, as follows:
+
+~~~
+agg_share = VDAF.out_shares_to_agg_share(agg_param, out_shares)
+~~~
+
+Note that for most VDAFs, it is possible to aggregate output shares as they
+arrive rather than wait until the batch is collected. To do so however, it is
+necessary to enforce the batch parameters as described in
+{{batch-parameter-validation}} so that the aggregator knows which aggregate
+share to update.
+
+The helper then encrypts `agg_share` under the collector's HPKE public
+key as described in {{aggregate-share-encrypt}}, yielding `encrypted_agg_share`.
+Encryption prevents the leader from learning the actual result, as it only has
+its own aggregate share and cannot compute the helper's.
+
+The helper responds to the leader with HTTP status 200 OK and a body consisting
+of an `AggregateShareResp`:
+
+~~~
+struct {
+    HpkeCiphertext encrypted_aggregate_share;
+} AggregateShareResp;
+~~~
+
+`encrypted_aggregate_share.config_id` is set to the collector's HPKE config ID.
+`encrypted_aggregate_share.enc` is set to the encapsulated HPKE context `enc`
+computed above and `encrypted_aggregate_share.ciphertext` is the ciphertext
+`encrypted_agg_share` computed above.
+
+After receiving the helper's response, the leader uses the HpkeCiphertext to
+respond to a collect request (see {{collect-flow}}).
+
+The leader MAY make multiple aggregate-share requests for a given batch interval
+and aggregation parameter and MUST get the same result each time.
```

Comment:
I don't understand why this behaviour is useful. Is this new, or is this already in the spec?

Reply:
I'm not sure if this is what Chris W. had in mind but I agree with this text. Suppose we have a task with `max_batch_lifetime = 1`. Consider:

1. Leader makes an `AggregateShareReq` for some batch interval.
2. Helper services the request, and marks that batch interval as having been collected once.
3. Helper attempts to transmit an `AggregateShareResp` to leader, but the message is truncated.

Now, if the leader retries its `AggregateShareReq`, the helper will refuse the request because the batch interval's lifetime has been consumed. So, to allow the leader to retry this request, the helper has to be willing to resend previously computed aggregate shares.

Reply:
Yanking this out to discuss in a separate issue: https://github.com/ietf-wg-ppm/ppm-specification/issues/226